xconq.man
.TH XCONQ 6 Utah
.SH NAME
xconq - X-based strategy game
.SH SYNOPSIS
.B
xconq
[ options ] ...
.SH DESCRIPTION
\fIXconq\fP is a multi-player strategic wargame.
Your goal is to save the world from the evil empires attempting to
take over.  You get cities, armies, airplanes, and ships with which to
accomplish your task.  Success is achieved by the neutralization or
defeat of the other sides, by any means available.  First you have to
find them!
.PP
Although the game can get complicated, to start you need only
know about the help command '?', which is valid at any time.
You start with only one city, which makes things even simpler.
The city will automatically produce a military unit, which you
will be prompted to move around.  Eventually, you will get more and
can attempt to take over other cities.
The game keeps going
until only one side is left in action;
this player is then declared the winner.
.PP
By default, \fIxconq\fP
starts up with one human (on the display in $DISPLAY) and one
machine, playing on a randomly generated 60x30 world.  Options are
available to set numbers and kinds of players, as well as the
kind of map used.  In addition, \fIxconq\fP has extensive and
elaborate facilities for building maps, scenarios, and historical
periods, resulting in a wide variety of games.
.SH OPTIONS
.TP 5
.I host:display
adds a human player to the game and assigns to the given host and display.
Players get created in the same order as they appear on the command line,
interleaved with the players created via \fB-e\fP and \fB-A\fP.
All of the player adding options may be used any number of times in any order.
.TP 5
.B \-A \fIhost:display\fP
attaches a machine player to the given display.  Ideal for those who like
to watch.
.TP 5
.B \-C
allows the machine player to play in cheat mode.  It will get a
complete update of the screen every turn.
.TP 5
.B \-e \fInumber\fP
sets the number of machine players not attached to displays.
Non-displayed machine players form alliance blocks of 5-9 countries.
.TP 5
.B \-m \fIname\fP
reads the map named \fIname\fP and starts up a game on it.
.TP 5
.B \-M \fIwidth height\fP
generates a random map of the given size.  The size must be at least 5x5,
although some periods will impose additional constraints on the lower bound.
In theory, there is no upper bound (but 200x100 is huge).
.TP 5
.B \-p \fIname\fP
reads the historical period specified by \fIname\fP, and sets everything
up for that period.
.TP 5
.B \-r
resets the list of players on the command line.  It is most useful when
placed before other player specifications, to clear out the default
players.
.TP 5
.B \-s \fIname\fP
reads a scenario with the name \fIname\fP.  A scenario sets up a particular
situation; it has a fixed number of sides that can participate.
.TP 5
.B \-t \fInumber\fP
toggles a chess-clock-like feature that limits the total time of play for
each side.  The amount is \fInumber\fP minutes per side.  Time is only counted
while actually waiting for input.
.TP 5
.B \-T \fInumber\fP
toggles a timeout that limits the length of each turn for all sides.
The amount is \fInumber\fP minutes per turn.  Time is counted
continuously.  The clock will stop updating after you have moved all
possible units, but if you then want to see how much time is left for
the other players, press ^L.
.TP 5
.B \-T \fInumber\fP
toggles a chess-clock-like feature that limits the per-round time of play for
each side.  The amount is \fInumber\fP minutes per side.
.TP 5
.B \-v
make the entire world seen by all players at the outset.
This is useful if exploration
is deemed to be time-consuming, or if the map is already known to everybody.
Some maps already have this enabled.
.\" .TP 5
.\" .B \-x
.\" allows scenarios, maps, periods, and players to be set up interactively
.\" via menus.  The menus should be self-explanatory.
.LP
In addition, \fB-B\fP and \fB-D\fP are recognized;
see the customization document for more details.
.SH EXAMPLES
.TP 8
.B
xconq
One human on local display, one machine, 60x30 random map
.TP 8
.B
xconq -m crater
One human vs one machine, on the "crater lake" map
.TP 8
.B
xconq -e 0 blitzen:0 kzin.eel:0
Three humans.  One on the local display, one on blitzen:0 in the same
network, and another: kzin.eel:0 in another network.  60x30 random map
.TP 8
.B
xconq -e 2 -M 90 40 -A shark:0
One human, three machines (one on display shark:0), 90x40 random map
.TP 8
.B
xconq -r -e 3 -m 1987 wind:0 surfer:0 -A springs:0
Six players (4 machine, 1 on springs:0; 2 human, on wind:0 and surfer:0),
all playing on a 360x122 map of the earth with present-day cities.  Major!
.SH "X DEFAULTS"
.TP 10
.B Class / name
You may specify these resources in your .Xres (or .Xdefaults) using a
spec of the form
.sp
    XConq.Class: value
.sp 
or
.sp
    XConq.name: value
.sp
and you can mix classes and names in the spec.  This is just standard X
resource stuff.
.PP
The following resources have a prefix of
.B XConq.Map.Period / xconq.map.%s
\.
.TP 10
.B HexDisplayMode / hexDisplayMode
This is how \fIxconq\fP will display the terrain hexes.  Possible values
are \fBFullHex\fP, \fBBorderHex\fP, \fBTerrIcons\fP, and \fBBothIcons\fP.
.TP 10
.B Background / ownColor
The color for your own units.
.TP 10
.B Background / alternateColor
I forgot
.TP 10
.B Background / differentColor
Reminders would be appreciated
.TP 10
.B Foreground / borderColor
It shouldn't be hard to figure out anyway
.TP 10
.B Foreground / grayColor
but my boss is looking over my shoulder
.TP 10
.B Foreground / enemyColor
The color of enemy units
.TP 10
.B Foreground / neutralColor
The color of neutral units
.TP 10
.B Foreground / goodColor
The color of allied units
.TP 10
.B Foreground / badColor
And I really shouldn't play this game until I've finished the project
.TP 10
.B Font / textFont
Font for all text.
.TP 10
.B Font / iconFont
Font for miscellaneous pictures.
.TP 10
.B Font/ helpFont
Font for the help pages.
.SH AUTHORS
Stan Shebs (shebs@cs.utah.edu)
.sp
Version 5.4 by Greg Fisher (fisher@cs.rutgers.edu)
.sp
Version 5.5 by Robert Forsman (thoth@lightning.cis.ufl.edu)
.sp
Version 6.0 in the works.  It doesn't work too well yet, Stan's still
fixing things.
.SH FILES
\fIsave.xconq\fP	saved game
.br
\fIemergency.save.xconq\fP	game saved after program crash
.br
\fIstats.xconq\fP	performance statistics for a game
.br
\fIcmds.xconq\fP	listing of commands
.br
\fIparms.xconq\fP	listing of units and their characteristics
.br
\fIview.xconq\fP	printable version of the map display
.br
\fI/local/lib/x11r4/xconq/*.map\fP	 predefined maps
.br
\fI/local/lib/x11r4/xconq/*.scn\fP	 predefined scenarios
.br
\fI/local/lib/x11r4/xconq/*.per\fP	 predefined historical periods
.br
\fI/local/lib/x11r4/xconq/*.onx\fP	 X10 fonts
.br
\fI/local/lib/x11r4/xconq/*.bdf\fP	 X11 uncompiled fonts
.br
\fI/local/lib/x11r4/xconq/*.snf\fP	 X11 compiled fonts
.br
\fI/local/lib/x11r4/xconq/xconq.news\fP	 news about features/additions
.SH "SEE ALSO"
.br
.IR "Liberating the World (Made Simple)"
.br
.IR "Customizing Xconq (Made Somewhat Simple)"
.SH DIAGNOSTICS
If a given display cannot be opened due to a missing \fIxhost(1)\fP command,
\fIxconq\fP will quit with a message "Error opening display host:0".
.PP
If the icon fonts cannot be found or opened, text will be substituted.
This is usually pretty yucky.
.PP
If the map is too small for the desired number of players, \fIxconq\fP will
complain about not being able to place units.
.PP
There are some panic messages which usually precede a core dump by a few
milliseconds.
.PP
Incorrect combinations of periods, maps, and scenarios
can generate a host of different error messages.
.SH BUGS
.PP
Since some periods/maps/scenarios are interdependent, they must be specified
in a particular order on the command line (period first usually).
.PP
Does not cope gracefully (it doesn't really cope at all) with
insufficient memory.
.PP
Some annoying behaviors are actually features.
.PP
Feel free to report bugs to xconq@uunet.uu.net.  There are people
there willing (if not able) to help you.  If you wish to listen in on
the list you may join by sending mail to xconq-request@uunet.uu.net.


CHANGES
Complete List of Changes From 5.1

(Note: this change list is cobbled together from various sources, and only
lightly edited.)

*** 5.5

Updated README

Moved all system-dependent defines from config.h to the Imakefile.
By virtue of the Imakefile, pcf font support was added as well as
several install & clean definitions.

Created an Imakefile.

* b4

Added marginal support for HPUX.  Including a user-written version of
strcasecmp.

Changed occurrences of #endif WORD to #endif /* WORD */ to shut up
some wimpy preprocessors.

Disabled the stack trace facility since the developer has a real
debugger and the signal interceptor prevents the core from recording
useful information.

Fixed the OpenWindows input focus problem (ICCCM non-compliance on our
part).

Enabled the game to support more than 7 players.

Added letters to the terrain font so that players above 9 would have
unique identifiers.

Added a little error checking to the resignation command.

Added a world-map magnification option.

Added the ability to add a player in the middle of the game.  Very
limited, and kinda screwy still.

Improved the internal geometry manager to handle periods with lots of
units a little better.

Since having 20 allied robots results in a lot of view updating, robot
alliances are now in blocks of 5-9 countries.  This will make it a
little easier for humans since robots can now fight each other.

fixed the do_look command so that it's not shadowed by the do_fill
command.

Added a little error checking to the map loader so that 0-length files
wouldn't crash it.

Expanded unit ids to ints.  This may cause problems since I can't
guarantee I've found all the code that will be affected.

Nailed a couple of bizzare conditions that CodeCenter found in the
machine player.

Bulletproofed the wake_unit code when passed an other==NULL.

Did a little tapdance to prevent the "builder has moved" message when
garrisoning.  Also fixed the "nonmover has moved" message during
garrisoning.  Also don't print message when garrison operation doesn't
destroy anything.

Improved the resource distribution system.  It seems to work well with
the standard period and 4tech.  Will it work with other periods as
well?

Tried to define the direction field in sergeant.c as a signed char,
ran into a few problems with Sun's sucky compiler.

Fixed bug where captured units regained full movement.  They now
merely hang onto whatever movement they had if the capturemoves
periodflag is enabled.

Added a few things to the manual page.  Documented the resources.

* b3

Better messages when searching for app-defaults and .Xdefaults.

area-radius reimplemented finally.

fixed problem in do_embark() caused by transport variable not being
initialized.

Made the help window behave better when handed unimplemented keystrokes.

Made saved game restore more robust and verbose in the face of errors.

Fixed version mismatch error message.

Removed wander_to() from movement routine until we can decide what the
movement routine's behaviour should be.

Fixed cost estimation routine in sergeant.c that was broken in the
shift from b1 to b2.

Allowed units to consume their last fuel with a mouse click movement.

Fixed problem where a maker intended to garrison its occupant would have
both itself and that occupant destroyed before the occupant got placed
correctly.

*b2

Added period.leavemap and period.repairscale to per2c output.

Fixed inconsistency in handling of repair for different repair rates
and scales.

Fixed curses display of neutral units to show a quote instead of "7".

Hacked period reader so that a non-zero extension in the header means
to not clear the period before reading (makes period inclusion more
useful :-) ).

Units not waiting to fill up will not be woken up when they become full.

Added terrain name information to the info window.

Fixed units changing sides so that you can't capture or give away
units that don't change side.  Also fixed it so that allied units
couldn't be given away by an ally.

Added X app-defaults handling (I think).

Added an icon hint to the main window, made the help window part of
main's windowgroup.

Improved X resource handling some: added ability to set the showmode
through resources.

Improved X resize and expose handling (the latter was a VERY dramatic
improvement).

Fixed problems with the worldmap viewport box not being drawn
correctly.

Added better event handling so that the help window could stay up
while the game was being played.  Also bound a few more keys in the
help window.

Added the move-toward-unit command, accessible with the ^M key (move
toward marked unit with confirmation) or <Shift><Mouse>.  The
possibility of binding different things to different mouse keys and
different modifiers does not escape me.

Added the ability to select the next unit to move in survey.

Added automatic delay of off-screen units so that the action does not
skip around as much (fewer viewport changes).

Nailed a SEGV in maybe_wakeup.

Added a parser keyword for the leave-map period variable.

Added a new keyword repair-scale that sets how many hps are repaired
each turn (assuming the chance succeeds).  The repair system needs to
be revamped in 6.0.

Fixed the border condition in handling of crippled units.

Added -d[display] to specify a display for the first (default) player.

Changed xconq library dir to be settable from an env variable XCONQLIB.

Stopped incessant scrolling when a unit is on the top edge.


xconq.c
/* Copyright (c) 1987, 1988  Stanley T. Shebs. */
/* This program may be used, copied, modified, and redistributed freely */
/* for noncommercial purposes, so long as this notice remains intact. */

/* Unlike most multi-player network programs, xconq does not use multiple */
/* processes.  Instead, it relies on the network capability inherent in  */
/* some graphical interfaces (such as X) to open up multiple displays. */

#include "config.h"
#include "misc.h"
#include "period.h"
#include "side.h"
#include "unit.h"
#include "map.h"
#include "global.h"
#include "version.h"

extern int nummaps;     /* needed to initialize number of maps read */

/* Definitions of globally global variables. */

Global global;          /* important (i.e. saved/restored) globals */

bool givenseen = FALSE; /* true if world known */
bool Debug = FALSE;     /* the debugging flag itself. */
bool Build = FALSE;     /* magic flag for scenario-builders */
bool Freeze = FALSE;    /* keeps machine players from moving */
int Cheat = 0;          /* If true allows machine to cheat */
bool No_give_r_and_d = FALSE; /* true if r&d cannot be given away XXX */
bool No_give_units = FALSE; /* true if units cannot be given away XXX */
bool Sequential = FALSE; /* true if sides move one at a time XXX */
bool humans[MAXSIDES];  /* flags for which players are human or machine */
/*** (UK) insert -> ***/
bool PartialGame=FALSE;  /* only a subset of players is playing */
bool ServerFlag=FALSE;   /* If true allows display requests */
int server_port=SERVER_PORT; /* port number for server facility */
/*** <- insert ***/ 

/* mxconq: Only current display initialized if set. */
bool singlePlayerMode = FALSE;


char *programname;      /* name of the main program */
char *xconqlib;		/* directory of xconq data files */
char *rawfilenames[MAXLOADED];  /* names specified on cmd line */
char version[] = VERSION;  /* version string */
char *hosts[MAXSIDES];  /* names of displays for each side */
char spbuf[BUFSIZE];    /* main printing buffer */
char tmpbuf[BUFSIZE];   /* a temp buffer */
/*** (UK) insert -> ***/
char tmpbuf2[BUFSIZE];   /* a temp buffer */
/*** <- insert ***/ 

int numfiles = 0;       /* number of data files asked for */
int numgivens = 0;      /* number of sides given on cmd line */
int numhumans = 0;      /* number of bona-fide human players */
int givenwidth = DEFAULTWIDTH;    /* requested width for a random map */
int givenheight = DEFAULTHEIGHT;  /* requested height for a random map */
/* XXX */
char nextturnexecfile[BUFSIZE]; /* execute file at beginning of each turn */

/* Indices of groups of mapfile types mentioned in collection of mapfiles. */


Side *winner = NULL;    /* the winning side (allies are also winners) */

/* Where it all begins... the main program's primary duty is command line */
/* interpretation, it hands off for everything else. */

main(argc, argv)
int argc;
char *argv[];
{
    bool eopt = FALSE;
    char ch;
    int i, numenemies;

    enter_procedure("Main");
    programname = argv[0];
    if ((xconqlib = getenv("XCONQLIB")) == NULL)
	xconqlib = XCONQLIB;
    nummaps = 0;
    nextturnexecfile[0] = '\0'; /* XXX */
    for (i = 0; i < MAXLOADED; ++i) rawfilenames[i] = "";
    global.giventime = global.timeout = 0;
    add_default_player();
    for (i = 1; i < argc; ++i) {
	if ((argv[i])[0] == '-') {
	    ch = (argv[i])[1];
	    if (Debug) printf("-%c\n", ch);
	    switch (ch) {
	    case 'A':
		if (i+1 >= argc) usage_error();
		add_player(FALSE, argv[++i]);
		break;
	    case 'B':
		Build = TRUE;
		Freeze = TRUE;
		break;
	    case 'C':
		Cheat = TRUE;
		break;
	    case 'd':  /* subsumes "-display" */
		if (i+1 >= argc) usage_error();
		hosts[0] = copy_string(argv[++i]);
		break;
	    case 'D':
		Debug++;
		break;
	    case 'e':
		if (i+1 >= argc) usage_error();
		eopt = TRUE;
		numenemies = atoi(argv[++i]);
		while (numenemies-- > 0) add_player(FALSE, (char *) NULL);
		break;
            case 'f':			/* XXX */
                if (i+1 >= argc) usage_error();
                strcpy(nextturnexecfile, argv[++i]);
                break;
	    case 'g':
		/* do -geometry for this player */
		break;
            case 'G':			/* XXX */
                No_give_units = TRUE;
                break;
	    case 'm':
		if (i+1 >= argc) usage_error();
		make_pathname((char *) NULL, argv[++i], "map", spbuf);
		rawfilenames[numfiles++] = copy_string(spbuf);
		break;
	    case 'M':	
		if (i+2 >= argc) usage_error();
		givenwidth = atoi(argv[++i]);
		givenheight = atoi(argv[++i]);
		break;
	    case 'p':
		if (i+1 >= argc) usage_error();
		make_pathname((char *) NULL, argv[++i], "per", spbuf);
		rawfilenames[numfiles++] = copy_string(spbuf);
		break;
	    case 'r':
		if (numgivens > 1) {
		    fprintf(stderr, "Warning: -r is resetting the list of\n");
		    fprintf(stderr, "players already given in the command.\n");
		}
		numgivens = numhumans = 0;
		break;
            case 'R':			/* XXX */
                No_give_r_and_d = TRUE;
                break;
	    case 's':
		if (i+1 >= argc) usage_error();
		make_pathname((char *) NULL, argv[++i], "scn", spbuf);
		rawfilenames[numfiles++] = copy_string(spbuf);
		break;
            case 'S':			/* XXX */
                Sequential = TRUE;
                break;
	    case 't':
		if (i+1 >= argc) usage_error();
		/* Converting to seconds for internal use */
		global.giventime = 60 * atoi(argv[++i]);
		if (!global.giventime) usage_error();
		break;
	    case 'T':
		if (i+1 >= argc) usage_error();
		/* Converting to seconds for internal use */
		global.timeout = 60 * atoi(argv[++i]);
		/* To keep clock from looking funny on startup: */
		global.starttime = time(0);
		break;
	    case 'v':
		givenseen = TRUE;
		break;
            case 'P':
                singlePlayerMode = TRUE;
                numgivens = numhumans = 0; /* same as -r */
                break;
/*** (UK) insert -> ***/
            case 'Q':
                PartialGame = TRUE;
		eopt=TRUE;
                numgivens = numhumans = 0; /* same as -r */
                break;
            case 'x':
                ServerFlag = TRUE;
                PartialGame = TRUE;
		eopt=TRUE;
                numgivens = numhumans = 0; /* same as -r */
                break;
            case 'X':
		if (i+1 >= argc) usage_error();
		server_port=atoi(argv[++i]);
                ServerFlag = TRUE;
                PartialGame = TRUE;
		eopt=TRUE;
                numgivens = numhumans = 0; /* same as -r */
                break;
/*** <- insert ***/ 
	    default:
		usage_error();
	    }
	} else {
	    /* We just have a host name for a human player */
	    add_player(TRUE, argv[i]);
	}
    }
    /* Put in a single machine opponent if no -e and no human opponents */
    if (!eopt && numgivens == 1) add_player(FALSE, (char *) NULL);
    /* Say hi to the user */
    printf("\n              Welcome to XCONQ version %s\n\n", version);
    maybe_dump_news();
   
    /* While away the hours setting everything up */
    init_random(-1);
    init_sighandlers();
    init_sides();
    init_units();
    init_game();
    init_displays1();
    init_mplayers();
    init_displays2();
    /* Relatively low chance of screwup now, so safe to delete saved game */
    /* if (saved_game()) remove_saved_game(); */
    /* Play! */
    play();
    return 0;
}

/* Complain and leave if command is garbage. */

usage_error()
{
    fprintf(stderr, "Usage: %s [display] [-A display] [-B] [-C] [-e n] [-m name]\n",
	    programname);
    fprintf(stderr, "  [-M w h] [-p name] [-r] [-s name] [-t n] [-T n] [-v]\n");
    fprintf(stderr, "  [-G] [-R] [-f filename]\n");
/*** (UK) insert -> ***/
    fprintf(stderr, "  [-P] [-Q] [-x] [-X port]\n");
/*** <- insert ***/
    exit(1);
}

/* Add a player into the array of players, making sure of no overflow. */
/* Fail if so, players probably made a typo on command line - help them */
/* out with a list of what players had been included. */

add_player(ahuman, host)
bool ahuman;
char *host;
{
    if (numgivens >= MAXSIDES) {
	fprintf(stderr, "At most %d player sides allowed!\n", MAXSIDES);
	fprintf(stderr, "(Valid ones were ");
	list_players(stderr);
	fprintf(stderr, ")\n");
	exit(1);
    }
    hosts[numgivens] = host;
    humans[numgivens] = ahuman;
    numgivens++;
    if (ahuman) numhumans++;
    if (Debug) printf("Added %s player (%s)\n",
		      (ahuman ? "human" : "machine"), (host ? host : ""));
}

/* The main loop that cycles through the turns and the phases. */
/* Always departs through game end phase (or by core dump :-( ). */

play()
{
/*** (HW) reinsert -> ***/
  int u;
  Side *side;
/*** <- reinsert ***/

    while (TRUE) {
	if (!midturnrestore) {
	  init_turn_phase();
	  spy_phase();
	  disaster_phase();
	  build_phase();
	  flush_dead_units();
	  supply_phase();
	  sort_units(FALSE);
	  auto_save_phase(); /* XXX */
	}
/*** (HW) reinsert -> ***/
	for_all_sides(side) for_all_unit_types(u) update_state(side, u);
        test_input();
/*** <- reinsert ***/
	movement_phase();
	consumption_phase();
	flush_dead_units();
	game_end_phase();
	if (numhumans == 0)
	  exit_robot_check();
    }
}

/* Do random small things to get the turn started, including resetting */
/* some side vars that could get things confused, if set wrong. */
/* Most movement-related things should get set later, at beginning of */
/* move phase. */

init_turn_phase()
{
    Side *side;

    ++global.time;
    if (Debug) printf("##### TURN %d #####\n", global.time);
    for_all_sides(side) {
	show_timemode(side,TRUE);
	update_clock(side);
	flush_input(side);
	flush_output(side);
    }
}

/* XXX perform an auto_save if the turn number is correct */

auto_save_phase()
{
        char filename[80];

        if (global.time % 5 != 0)
                return;
        sprintf(filename, "%s.auto.%d", SAVEFILE, global.time);
        if (!write_savefile(filename))
            notify_all("Can't open file \"%s\"!", filename);
}

/* The win/lose phase assesses the sides' performance during that turn, */
/* and can end the game.  In fact, it is the only way to get out, except */
/* for the quick exit command.  With multiple mixed players, there are a */
/* number of possibilities at game end, such as an all machine win.  Also, */
/* it is possible for all players to lose simultaneously! */

game_end_phase()
{
    bool enemiesleft = FALSE;
    int sidesleft = 0, testrslt;
    Side *side, *side2, *survivor;

    if (Debug) printf("Entering game end phase\n");
    /* See if any sides have won or lost */
    if (global.time >= global.endtime) {
	notify_all("The end of the world has arrived!");
	for_all_sides(side) side_loses(side, (Side *) NULL);
    } else if (Build) {
	/* No winning or losing while building */
    } else {
	for_all_sides(side) {
	    if (!side->lost) {
		if (side->timedout) {
		    notify_all("The %s ran out of time!",
			       plural_form(side->name));
		    side_loses(side, (Side *) NULL);
		} else if (all_units_gone(side)) {
		    notify_all("The %s have been wiped out!",
			       plural_form(side->name));
		    side_loses(side, (Side *) NULL);
		} else {
		    testrslt = condition_true(side);
		    if (testrslt > 0) side_wins(side);
		    if (testrslt < 0) side_loses(side, (Side *) NULL);
		}
	    }
	}
    }
    /* See if any opposing sides left */
    for_all_sides(side) {
	for_all_sides(side2) {
	    if (!side->lost && !side2->lost && enemy_side(side, side2))
		enemiesleft = TRUE;
	}
	if (!side->lost) {
	    survivor = side;
	    sidesleft++;
	}
    }
    /* Decide if the game is over */
    if (numsides == 1) {
	/* A one-player game can't end */
    } else if (sidesleft == 0) {
        exit_xconq();
    } else if (!enemiesleft) {
	winner = survivor;
	if (sidesleft == 1) {
	    notify(winner, "You have prevailed over your enemies!");
	} else {
	    for_all_sides(side) {
		if (!side->lost) {
		    notify(side, "Your alliance has defeated its enemies!");
		}
	    }
	}
	wait_to_exit(winner);
	exit_xconq();
    } else {
        /* Game isn't over yet */
    }
}

/* Quicky test for absence of all units. */

all_units_gone(side)
Side *side;
{
    int u;

    for_all_unit_types(u) if (side->units[u] > 0) return FALSE;
    return TRUE;
}

has_type(unit, t)
Unit *unit;
int   t;
{
    Unit *occ;

    if (t == -1 || unit->type == t) return TRUE;

    for_all_occupants(unit, occ)
	if (alive(occ) && occ->type == t)
	    return TRUE;

    return FALSE;
}

/* Check out all the winning and losing conditions, returning a flag */
/* on which, if any, are true.  The three return values are 1 for win, */
/* -1 for lose, and 0 for undecided. */

condition_true(side)
Side *side;
{
    int i, u, r, amt = 0, is_here = 0;
    Unit *unit, *occ;
    Condition *cond;

    for (i = 0; i < global.numconds; ++i) {
	cond = &(global.winlose[i]);
	if ((between(cond->starttime, global.time, cond->endtime)) &&
	    (cond->sidesn == -1 || cond->sidesn == side_number(side))) {
	    switch (cond->type) {
	    case TERRITORY:
		for_all_unit_types(u) {
		    amt += side->units[u] * utypes[u].territory;
		}
		if (cond->win) {
		    if (amt >= cond->n) {
			notify_all("The %s have enough territory to win!",
				   plural_form(side->name));
			return 1;
		    }
		} else {
		    if (amt < cond->n) {
			notify_all("The %s don't have enough territory!",
				   plural_form(side->name));
			return -1;
		    }
		}
		break;
	    case UNITCOUNT:
		if (cond->win) {
		    for_all_unit_types(u) {
			if (side->units[u] < cond->units[u]) return 0;
		    }
		    notify_all("The %s have enough units to win!",
			       plural_form(side->name));
		    return 1;
		} else {
		    for_all_unit_types(u) {
			if (side->units[u] > cond->units[u]) return 0;
		    }
		    notify_all("The %s don't have the required units!",
			       plural_form(side->name));
		    return -1;
		}
		break;
	    case RESOURCECOUNT:
		if (cond->win) {
		    for_all_resource_types(r) {
			if (side->resources[r] < cond->resources[r]) return 0;
		    }
		    notify_all("The %s have enough resources to win!",
			       plural_form(side->name));
		    return 1;
		} else {
		    for_all_resource_types(r) {
			if (side->resources[r] > cond->resources[r]) return 0;
		    }
		    notify_all("The %s don't have the required resources!",
			       plural_form(side->name));
		    return -1;
		}
		break;
	    case POSSESSION:
		unit = unit_at(cond->x, cond->y);
		if (cond->win) {
		    if (unit != NULL && unit->side == side &&
			has_type(unit, cond->n)) {
			notify_all("The %s have a %s in hex %d,%d!",
				   plural_form(side->name),
				   utypes[cond->n].name,
				   cond->x, cond->y);
			return 1;
		    }
		} else {
		    if (unit == NULL || unit->side != side ||
			!has_type(unit, cond->n)) {
			notify_all("The %s don't have a %s in hex %d,%d!",
				   plural_form(side->name),
				   utypes[cond->n].name,
				   cond->x, cond->y);
			return -1;
		    }
		}
		break;
	    default:
		case_panic("condition type", cond->type);
		break;
	    }
	}
    }
    return 0;
}

/* Winning is defined as not losing :-) */

side_wins(side)
Side *side;
{
    Side *side2;

    for_all_sides(side2) {
	if (!allied_side(side, side2)) side_loses(side2, (Side *) NULL);
    }
}

/* Be rude to losers and give all their units to their defeaters (might */
/* not be one, so units become neutral or dead).  Give the loser a chance */
/* to study the screen, mark the side as lost, then pass on this detail to */
/* all the other sides and remove spurious leftover images of units. */

side_loses(side, victor)
Side *side, *victor;
{
    int ux, uy;
    Unit *unit;
    Side *side2;

    notify(side, "You lost, sucker");
    if (active_display(side)) {
	wait_to_exit(side);
	close_display(side);
    }
    side->humanp = FALSE; /* just make sure */
    while (side->unithead->next != side->unithead) {
      unit = side->unithead->next;
      if (alive(unit)) {
	ux = unit->x;  uy = unit->y;
	unit_changes_side(unit, victor, CAPTURE, SURRENDER);
	all_see_hex(ux, uy);
      } else {
	/* units could die instead of changing sides.   Inefficient if */
	/* it happens often, but sides don't resign often.  We need to */
	/* flush the dead units to make sure that the side has no */
	/* units.  Easier this way, since unit_changes_side could get */
	/* a lot of occupants, so we can't just step through the list. */
	flush_dead_units();
      }
    }
    side->lost = TRUE;
    if (nextturnexecfile[0] != '\0') { /* XXX */
        char buf[BUFSIZE];

        sprintf(buf, "%s side %d %d %d", nextturnexecfile, side_number(side),
                (side->lost ? 0 : 1), (side->more_units ? 1 : 0));
        system(buf);
    }
    for_all_sides(side2) {
	if (active_display(side2)) {
	    remove_images(side2, side_number(side));
	    if (side != side2) show_all_sides(side2,TRUE);
	}
    }
}

/* Leave screen up so everybody can study it, and allow any input to take */
/* it down.  OK to freeze all the other sides at once here. */

wait_to_exit(side)
Side *side;
{
    if (humanside(side) && active_display(side)) {
	sprintf(side->promptbuf, "[Do anything to exit]");
	show_prompt(side);
	freeze_wait(side);
    }
}

/* This routine should be called before any sort of non-death exit. */
/* Among other things, it updates the statistics. */

exit_xconq()
{
    int n = 0;
    Unit *unit;
    Side *side;

    close_displays();
    /* Need to kill off all units to finish up statistics */
    for_all_units(side, unit)
      kill_unit(unit, (winner ? VICTOR : SURRENDER));
    print_statistics();
    /* Offer a one-line comment on the game and then leave */
    for_all_sides(side) if (!side->lost) n++;
    switch (n) {
    case 0:
	printf("\nEverybody lost!\n\n");
	break;
    case 1:
	printf("\nThe %s (%s) won!\n\n",
	       plural_form(winner->name),
	       (winner->host ? winner->host : "machine"));
	break;
    default:
	sprintf(spbuf, "");
	for_all_sides(side) {
	    if (!side->lost) {
		if (strlen(spbuf) > 0) {
		    sprintf(tmpbuf, "/");
		    strcat(spbuf, tmpbuf);
		}
		sprintf(tmpbuf, "%s", side->name);
		strcat(spbuf, tmpbuf);
		if (side->host) {
		    sprintf(tmpbuf, "(%s)", side->host);
		    strcat(spbuf, tmpbuf);
		}
	    }
	}
	printf("\nThe %s alliance won!\n\n", spbuf);
	break;
    }
    exit(0);
}

/* Shut down displays - should be done before any sort of exit. */

close_displays()
{
    Side *side;

/*** (HW) change -> ***/
    for_all_sides(side) {
      /* avoid drawing without display at the end of the game */
      side->show_orders = FALSE;
      side->showsorderutype = NOTHING;
      if (active_display(side)) close_display(side);
    }
/*** was:
    for_all_sides(side) if (active_display(side)) close_display(side);
*** <- change ***/
}

/* This is to find out how everybody did. */

print_statistics()
{
    Side *side;
    FILE *fp;

    enter_procedure("print stats");
#ifdef STATISTICS
    if ((fp = fopen(STATSFILE, "w")) != NULL) {
	for_all_sides(side) {
	    print_side_results(fp, side);
	    print_unit_record(fp, side);
	    print_combat_results(fp, side);
	    if (side->next != NULL) fprintf(fp, "\f\n");
	}
	fclose(fp);
    } else {
	fprintf(stderr, "Can't open statistics file \"%s\"\n", STATSFILE);
    }
#endif /* STATISTICS */
    exit_procedure();
}

/* List all the specified players briefly. */

list_players(fp)
FILE *fp;
{
    int i;

    if (numgivens > 0) {
	fprintf(fp, "%s", (hosts[0] ? hosts[0] : "machine"));
	for (i = 1; i < numgivens; ++i) {
	    fprintf(fp, ", %s", (hosts[i] ? hosts[i] : "machine"));
	}
    } else {
	fprintf(fp, "no players defined.");
    }
}
